File: brigade-medicale-frontend/src/app/features/triage/core/services/triage.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Observable, of, Subject } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { ConfigService } from '../../../../core/services/config.service';
import { ApiResponse } from '../../../../core/models/api-response.model';
import { TriageRecord, CreateTriageDto, TriageDraft } from '../models/triage.model';

interface QueuedTriage {
  id: string;
  data: CreateTriageDto;
  retries: number;
  lastRetry?: number;
}

/**
 * Service pour gérer les données de Triage (prise de constantes)
 * Endpoints: /api/triage
 * Fallback: localStorage pour mode offline avec queue de synchronisation
 */
@Injectable({
  providedIn: 'root'
})
export class TriageService {
  private readonly DRAFT_KEY = 'triageDraft';
  private readonly QUEUE_KEY = 'triageOfflineQueue';
  private readonly MAX_RETRIES = 3;
  private readonly SYNC_INTERVAL = 5000; // 5 secondes

  private syncSubject = new Subject<boolean>();
  public sync$ = this.syncSubject.asObservable();

  private isSyncing = false;
  private syncTimer: any;

  constructor(private http: HttpClient, private configService: ConfigService) {
    this.initOnlineMonitoring();
  }

  private getApiUrl(): string {
    return `${this.configService.getApiUrl()}/triage`;
  }

  /**
   * Initialise la surveillance du statut online/offline
   */
  private initOnlineMonitoring(): void {
    window.addEventListener('online', () => {
      console.log('✓ Connexion rétablie → Synchronisation de la queue');
      this.syncOfflineQueue();
    });
  }

  /**
   * Enregistre un nouveau triage
   * POST /api/triage
   * Si offline: ajoute à la queue et tente la synchronisation
   */
  createTriage(data: CreateTriageDto): Observable<ApiResponse<TriageRecord>> {
    return this.http.post<ApiResponse<TriageRecord>>(`${this.getApiUrl()}`, data)
      .pipe(
        catchError((error: HttpErrorResponse) => {
          // Gestion du mode offline (pas de connexion ou 404)
          if (error.status === 404 || !navigator.onLine) {
            console.warn('⚠️ Mode offline détecté → Triage ajouté à la queue');
            const queuedItem: QueuedTriage = {
              id: `triage-${Date.now()}`,
              data,
              retries: 0
            };
            this.addToQueue(queuedItem);

            // Tenter la synchronisation si online
            if (navigator.onLine) {
              this.syncOfflineQueue();
            }

            // Retourner une réponse fictive
            return of({
              success: true,
              message: '[Mode offline] Triage en attente de synchronisation',
              data: {
                id: queuedItem.id,
                patientId: data.patientId,
                recordedAt: new Date().toISOString(),
                temperature: data.temperature,
                systolicBP: data.systolicBP,
                diastolicBP: data.diastolicBP,
                pulse: data.pulse,
                weight: data.weight,
                height: data.height,
                spO2: data.spO2,
                complaint: data.complaint,
                notes: data.notes,
                urgencyLevel: data.urgencyLevel,
                status: 0
              } as any
            });
          }
          throw error;
        })
      );
  }

  /**
   * Ajoute un triage à la queue offline
   */
  private addToQueue(item: QueuedTriage): void {
    const queue = this.getQueue();
    queue.push(item);
    localStorage.setItem(this.QUEUE_KEY, JSON.stringify(queue));
    console.log(`✓ Triage #${item.id} ajouté à la queue (${queue.length} items)`);
  }

  /**
   * Récupère la queue offline
   */
  private getQueue(): QueuedTriage[] {
    const queue = localStorage.getItem(this.QUEUE_KEY);
    return queue ? JSON.parse(queue) : [];
  }

  /**
   * Synchronise la queue offline avec le serveur
   */
  public syncOfflineQueue(): void {
    if (this.isSyncing || !navigator.onLine) return;

    this.isSyncing = true;
    const queue = this.getQueue();

    if (queue.length === 0) {
      this.isSyncing = false;
      return;
    }

    console.log(`🔄 Synchronisation de la queue: ${queue.length} item(s)`);

    let completedRequests = 0;
    const totalRequests = queue.filter(item => item.retries < this.MAX_RETRIES).length;

    if (totalRequests === 0) {
      this.isSyncing = false;
      return;
    }

    queue.forEach(item => {
      if (item.retries < this.MAX_RETRIES) {
        this.http.post<ApiResponse<TriageRecord>>(`${this.getApiUrl()}`, item.data)
          .pipe(
            catchError(error => {
              item.retries++;
              item.lastRetry = Date.now();

              if (item.retries >= this.MAX_RETRIES) {
                console.error(`❌ Sync failed for ${item.id} after ${this.MAX_RETRIES} retries`);
              } else {
                console.warn(`⚠️ Sync failed for ${item.id}, retry ${item.retries}/${this.MAX_RETRIES}`);
              }

              return of(null);
            })
          )
          .subscribe(response => {
            if (response?.success) {
              console.log(`✓ Triage ${item.id} synchronized successfully`);
              this.removeFromQueue(item.id);
              this.syncSubject.next(true);
            } else {
              item.retries++;
              item.lastRetry = Date.now();
            }

            // Sauvegarder l'état de la queue
            const updatedQueue = this.getQueue();
            const itemIndex = updatedQueue.findIndex(q => q.id === item.id);
            if (itemIndex >= 0) {
              updatedQueue[itemIndex] = item;
              localStorage.setItem(this.QUEUE_KEY, JSON.stringify(updatedQueue));
            }

            completedRequests++;
            // Reset isSyncing only when all requests have completed
            if (completedRequests === totalRequests) {
              this.isSyncing = false;
            }
          });
      }
    });
  }

  /**
   * Supprime un triage de la queue
   */
  private removeFromQueue(id: string): void {
    let queue = this.getQueue();
    queue = queue.filter(item => item.id !== id);
    localStorage.setItem(this.QUEUE_KEY, JSON.stringify(queue));
    console.log(`✓ Triage ${id} removed from queue`);
  }

  /**
   * Obtient le nombre d'items en attente de synchronisation
   */
  getQueueLength(): number {
    return this.getQueue().length;
  }

  /**
   * Obtient le contenu complet de la queue
   */
  getQueueItems(): QueuedTriage[] {
    return this.getQueue();
  }

  /**
   * Récupère le triage le plus récent pour un patient
   * GET /api/triage/latest?patientId={id}
   */
  getLatestTriage(patientId: string): Observable<ApiResponse<TriageRecord>> {
    return this.http.get<ApiResponse<TriageRecord>>(`${this.getApiUrl()}/latest`, {
      params: { patientId }
    })
      .pipe(
        catchError((error: HttpErrorResponse) => {
          if (error.status === 404) {
            console.warn('⚠️ Endpoint /api/triage/latest absent (404) → Pas de triage trouvé');
            return of({
              success: false,
              message: 'Aucun triage trouvé',
              data: null as any
            });
          }
          throw error;
        })
      );
  }

  /**
   * Récupère le triage pour une visite spécifique
   * GET /api/triage/by-visit/{visitId}
   */
  getTriageByVisit(visitId: string): Observable<ApiResponse<TriageRecord>> {
    return this.http.get<ApiResponse<TriageRecord>>(`${this.getApiUrl()}/by-visit/${visitId}`)
      .pipe(
        catchError((error: HttpErrorResponse) => {
          if (error.status === 404) {
            console.warn(`⚠️ Aucun triage trouvé pour visite ${visitId}`);
            return of({
              success: false,
              message: 'Aucun triage pour cette visite',
              data: null as any
            });
          }
          throw error;
        })
      );
  }

  /**
   * Sauvegarde un brouillon de triage en localStorage (mode offline)
   */
  saveDraft(draft: TriageDraft): void {
    const draftWithTimestamp: TriageDraft = {
      ...draft,
      savedAt: Date.now()
    };
    localStorage.setItem(this.DRAFT_KEY, JSON.stringify(draftWithTimestamp));
    console.log('✓ Triage en brouillon sauvegardé localement');
  }

  /**
   * Récupère le brouillon sauvegardé
   */
  getDraft(): TriageDraft | null {
    const draft = localStorage.getItem(this.DRAFT_KEY);
    return draft ? JSON.parse(draft) : null;
  }

  /**
   * Efface le brouillon
   */
  clearDraft(): void {
    localStorage.removeItem(this.DRAFT_KEY);
    console.log('✓ Brouillon supprimé');
  }

  /**
   * Récupère les triages du jour
   * GET /api/triage/today
   */
  getTodayTriages(): Observable<ApiResponse<TriageRecord[]>> {
    return this.http.get<ApiResponse<TriageRecord[]>>(`${this.getApiUrl()}/today`)
      .pipe(
        catchError((error: HttpErrorResponse) => {
          if (error.status === 404) {
            console.warn('⚠️ Endpoint /api/triage/today absent (404) → Aucun triage trouvé');
            return of({
              success: true,
              message: 'Aucun triage trouvé pour aujourd\'hui',
              data: [] as any
            });
          }
          throw error;
        })
      );
  }

  /**
   * Calcul l'IMC
   */
  calculateIMC(weightKg: number, heightCm: number): number {
    const heightM = heightCm / 100;
    return Math.round((weightKg / (heightM * heightM)) * 10) / 10;
  }
}

File: src/BrigadeMedicale.Patient.Mobile/Features/Triage/Services/TriageService.cs
using System.Text.Json;
using BrigadeMedicale.Patient.Mobile.Core.Http;
using BrigadeMedicale.Patient.Mobile.Core.Models;

namespace BrigadeMedicale.Patient.Mobile.Features.Triage.Services;

/// <summary>
/// Service pour gérer les triages (prise de constantes)
/// Supporte le mode offline avec queue de synchronisation
/// </summary>
public interface ITriageService
{
    Task<ApiResponse<TriageRecordDto>> CreateTriageAsync(CreateTriageDto dto);
    Task<ApiResponse<TriageRecordDto>> GetLatestTriageAsync(Guid patientId);
    Task<List<TriageRecordDto>> GetTodayTriagesAsync();
    void SaveDraft(CreateTriageDto draft);
    CreateTriageDto? GetDraft();
    void ClearDraft();
    int GetQueueLength();
    Task SyncOfflineQueueAsync();
    double CalculateIMC(double weight, int height);
}

public class TriageService : ITriageService
{
    private readonly IApiClient _apiClient;
    private readonly string _apiUrl = "/triage";
    private readonly string _draftKey = "triageDraft";
    private readonly string _queueKey = "triageOfflineQueue";
    private const int MaxRetries = 3;

    private bool _isSyncing = false;
    private List<QueuedTriage> _offlineQueue = new();

    public TriageService(IApiClient apiClient)
    {
        _apiClient = apiClient;
        LoadQueueFromStorage();
    }

    /// <summary>
    /// Crée un nouveau triage
    /// </summary>
    public async Task<ApiResponse<TriageRecordDto>> CreateTriageAsync(CreateTriageDto dto)
    {
        try
        {
            var response = await _apiClient.PostAsync<TriageRecordDto>(_apiUrl, dto);

            if (response.Success)
            {
                ClearDraft();
                return response;
            }

            // Si offline, ajouter à la queue
            if (!IsOnline())
            {
                return HandleOfflineCreate(dto);
            }

            return response;
        }
        catch (Exception ex)
        {
            MainThread.BeginInvokeOnMainThread(() =>
            {
                Application.Current?.MainPage?.DisplayAlert("Erreur", $"Erreur lors du triage: {ex.Message}", "OK");
            });

            // En cas d'erreur, ajouter à la queue si offline
            if (!IsOnline())
            {
                return HandleOfflineCreate(dto);
            }

            throw;
        }
    }

    /// <summary>
    /// Récupère le dernier triage d'un patient
    /// </summary>
    public async Task<ApiResponse<TriageRecordDto>> GetLatestTriageAsync(Guid patientId)
    {
        try
        {
            var url = $"{_apiUrl}/latest?patientId={patientId}";
            return await _apiClient.GetAsync<TriageRecordDto>(url);
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"Error fetching latest triage: {ex.Message}");
            return new ApiResponse<TriageRecordDto>
            {
                Success = false,
                Message = "Impossible de récupérer le triage"
            };
        }
    }

    /// <summary>
    /// Récupère les triages du jour
    /// </summary>
    public async Task<List<TriageRecordDto>> GetTodayTriagesAsync()
    {
        try
        {
            var url = $"{_apiUrl}/today";
            var response = await _apiClient.GetAsync<List<TriageRecordDto>>(url);

            if (response.Success && response.Data != null)
            {
                return response.Data;
            }

            return new List<TriageRecordDto>();
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"Error fetching today's triages: {ex.Message}");
            return new List<TriageRecordDto>();
        }
    }

    /// <summary>
    /// Sauvegarde un brouillon de triage en local
    /// </summary>
    public void SaveDraft(CreateTriageDto draft)
    {
        try
        {
            var json = JsonSerializer.Serialize(draft);
            Preferences.Default.Set(_draftKey, json);
            Debug.WriteLine("✓ Triage en brouillon sauvegardé localement");
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"Error saving draft: {ex.Message}");
        }
    }

    /// <summary>
    /// Récupère le brouillon sauvegardé
    /// </summary>
    public CreateTriageDto? GetDraft()
    {
        try
        {
            if (Preferences.Default.ContainsKey(_draftKey))
            {
                var json = Preferences.Default.Get(_draftKey, string.Empty);
                if (!string.IsNullOrEmpty(json))
                {
                    return JsonSerializer.Deserialize<CreateTriageDto>(json);
                }
            }
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"Error loading draft: {ex.Message}");
        }
        return null;
    }

    /// <summary>
    /// Efface le brouillon
    /// </summary>
    public void ClearDraft()
    {
        try
        {
            Preferences.Default.Remove(_draftKey);
            Debug.WriteLine("✓ Brouillon supprimé");
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"Error clearing draft: {ex.Message}");
        }
    }

    /// <summary>
    /// Obtient le nombre d'items en attente de synchronisation
    /// </summary>
    public int GetQueueLength()
    {
        return _offlineQueue.Count;
    }

    /// <summary>
    /// Synchronise la queue offline avec le serveur
    /// </summary>
    public async Task SyncOfflineQueueAsync()
    {
        if (_isSyncing || IsOnline() == false || _offlineQueue.Count == 0)
            return;

        _isSyncing = true;

        try
        {
            var itemsToSync = _offlineQueue.Where(item => item.Retries < MaxRetries).ToList();
            int completedRequests = 0;
            int totalRequests = itemsToSync.Count;

            Debug.WriteLine($"🔄 Synchronisation de la queue: {totalRequests} item(s)");

            foreach (var item in itemsToSync)
            {
                try
                {
                    var response = await _apiClient.PostAsync<TriageRecordDto>(_apiUrl, item.Data);

                    if (response.Success)
                    {
                        Debug.WriteLine($"✓ Triage {item.Id} synchronized successfully");
                        _offlineQueue.Remove(item);
                    }
                    else
                    {
                        item.Retries++;
                        if (item.Retries >= MaxRetries)
                        {
                            Debug.WriteLine($"❌ Sync failed for {item.Id} after {MaxRetries} retries");
                        }
                    }
                }
                catch (Exception ex)
                {
                    item.Retries++;
                    Debug.WriteLine($"⚠️ Sync failed for {item.Id}: {ex.Message}, retry {item.Retries}/{MaxRetries}");
                }

                completedRequests++;
            }

            SaveQueueToStorage();
            Debug.WriteLine($"✓ Synchronisation terminée: {_offlineQueue.Count} items restants");
        }
        finally
        {
            _isSyncing = false;
        }
    }

    /// <summary>
    /// Calcule l'IMC
    /// </summary>
    public double CalculateIMC(double weight, int height)
    {
        if (height <= 0) return 0;
        double heightM = height / 100.0;
        return Math.Round((weight / (heightM * heightM)) * 10) / 10;
    }

    // ========== Private Methods ==========

    private bool IsOnline() => Connectivity.Current.NetworkAccess == NetworkAccess.Internet;

    private ApiResponse<TriageRecordDto> HandleOfflineCreate(CreateTriageDto dto)
    {
        var queuedItem = new QueuedTriage
        {
            Id = Guid.NewGuid().ToString(),
            Data = dto,
            Retries = 0
        };

        _offlineQueue.Add(queuedItem);
        SaveQueueToStorage();

        Debug.WriteLine($"⚠️ Mode offline détecté → Triage ajouté à la queue ({_offlineQueue.Count} items)");

        return new ApiResponse<TriageRecordDto>
        {
            Success = true,
            Message = "[Mode offline] Triage en attente de synchronisation",
            Data = new TriageRecordDto
            {
                Id = Guid.Parse(queuedItem.Id),
                PatientId = dto.PatientId,
                RecordedAt = DateTime.Now,
                Temperature = dto.Temperature,
                Systolic = dto.Systolic,
                Diastolic = dto.Diastolic,
                Pulse = dto.Pulse,
                Weight = dto.Weight,
                Height = dto.Height,
                SpO2 = dto.SpO2,
                IMC = CalculateIMC(dto.Weight, dto.Height),
                Complaint = dto.Complaint,
                Notes = dto.Notes,
                UrgencyLevel = dto.UrgencyLevel,
                Status = TriageStatus.Pending
            }
        };
    }

    private void SaveQueueToStorage()
    {
        try
        {
            var json = JsonSerializer.Serialize(_offlineQueue);
            Preferences.Default.Set(_queueKey, json);
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"Error saving queue: {ex.Message}");
        }
    }

    private void LoadQueueFromStorage()
    {
        try
        {
            if (Preferences.Default.ContainsKey(_queueKey))
            {
                var json = Preferences.Default.Get(_queueKey, string.Empty);
                if (!string.IsNullOrEmpty(json))
                {
                    _offlineQueue = JsonSerializer.Deserialize<List<QueuedTriage>>(json) ?? new();
                }
            }
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"Error loading queue: {ex.Message}");
        }
    }

    private class QueuedTriage
    {
        public string Id { get; set; } = string.Empty;
        public CreateTriageDto Data { get; set; } = new();
        public int Retries { get; set; }
        public DateTime LastRetry { get; set; }
    }
}
